build ruby linked against jemalloc

joshrwolf opened the pull request:
this changes all our ruby to be linked against `jemalloc`. for applications like webapps (rails) or multithreaded workloads (puma), we see significant memory usage savings compared to malloc.

I don't know enough as to whether we should do this globally, since this introduces some size overhead to our binaries, so I'm going to leave it to the experts to yay or nay this.

if nay, we'd likely need a jemalloc enabled subpackage or something similar just for the sheer perf gains we see against any rails/puma app

joshrwolf committed:
compile ruby linked against jemalloc

Signed-off-by: Josh Wolf <josh@wolfs.io>

dannf reviewed (changes requested):
I don't have any prior experience with jemalloc. But from what I can google, it seems like a good choice for x86, but I'm less sure about ARM:

https://github.com/facebook/buck2/pull/693

I'd support this if we guarded it as x86-only for now, but I think we'd want more investigation before turning it on for ARM.

xnox commented:
> I don't have any prior experience with jemalloc. But from what I can google, it seems like a good choice for x86, but I'm less sure about ARM:
> 
> [facebook/buck2#693](https://github.com/facebook/buck2/pull/693)
> 
> I'd support this if we guarded it as x86-only for now, but I think we'd want more investigation before turning it on for ARM.

We do this thing `export JEMALLOC_SYS_WITH_LG_PAGE=16` which supposedly what one needs to do to make it better on ARM. And I don't know if this is a build-time option only, or if one needs this at runtime too. So I think we transitioned to 16 by default a long time ago.

joshrwolf commented:
> I'd support this if we guarded it as x86-only for now, but I think we'd want more investigation before turning it on for ARM.

@dannf what other investigation would you like to see? and in the meantime, if we need this (both archs) for rails apps what would be a good temporary path forward to provide an apk with ruby-jemalloc?

xnox commented:
We have jemalloc enabled in more than 7 packages already; including on arm64; as we did so in:
- 2fe52b524ca7270e66f1742297d330bbc29c35b6

I would be more inclined to have it jemmaloc on, on both arches. The mentioned bug report is fixed with above change to build environment on arm64.

dannf commented:
> > I don't have any prior experience with jemalloc. But from what I can google, it seems like a good choice for x86, but I'm less sure about ARM:
> > [facebook/buck2#693](https://github.com/facebook/buck2/pull/693)
> > I'd support this if we guarded it as x86-only for now, but I think we'd want more investigation before turning it on for ARM.
> 
> We do this thing `export JEMALLOC_SYS_WITH_LG_PAGE=16` which supposedly what one needs to do to make it better on ARM. And I don't know if this is a build-time option only, or if one needs this at runtime too. So I think we transitioned to 16 by default a long time ago.

OK, that makes sense, but why is it 16 and not 64?

dannf reviewed (approved):
smoke tested ruby-3.4 on 64K and 4K kernels - seems to be fine :)

dannf reviewed (changes requested):
Actually, I noticed a failure running `make test ruby3.4-rails-8.0`:

```
2025/01/22 21:02:26 WARN current directory: /usr/lib/ruby/gems/3.4.0/gems/bcrypt_pbkdf-1.1.1/ext/mri
2025/01/22 21:02:26 WARN make DESTDIR\= sitearchdir\=./.gem.20250122-11-urkb5f
2025/01/22 21:02:26 WARN sitelibdir\=./.gem.20250122-11-urkb5f
2025/01/22 21:02:26 WARN aarch64-unknown-linux-gnu-gcc -I. -I/usr/include/ruby-3.4.0/aarch64-linux-gnu
2025/01/22 21:02:26 WARN -I/usr/include/ruby-3.4.0/ruby/backward -I/usr/include/ruby-3.4.0 -I.   -fPIC  
2025/01/22 21:02:26 WARN -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wdiv-by-zero
2025/01/22 21:02:26 WARN -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wpointer-arith
2025/01/22 21:02:26 WARN -Wwrite-strings -Wold-style-definition -Wimplicit-fallthrough=0
2025/01/22 21:02:26 WARN -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand
2025/01/22 21:02:26 WARN -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings
2025/01/22 21:02:26 WARN -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign
2025/01/22 21:02:26 WARN -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value
2025/01/22 21:02:26 WARN -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable
2025/01/22 21:02:26 WARN -Wmisleading-indentation -Wundef  -fPIC  -o bcrypt_pbkdf.o -c bcrypt_pbkdf.c
2025/01/22 21:02:26 WARN aarch64-unknown-linux-gnu-gcc -I. -I/usr/include/ruby-3.4.0/aarch64-linux-gnu
2025/01/22 21:02:26 WARN -I/usr/include/ruby-3.4.0/ruby/backward -I/usr/include/ruby-3.4.0 -I.   -fPIC  
2025/01/22 21:02:26 WARN -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wdiv-by-zero
2025/01/22 21:02:26 WARN -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wpointer-arith
2025/01/22 21:02:26 WARN -Wwrite-strings -Wold-style-definition -Wimplicit-fallthrough=0
2025/01/22 21:02:26 WARN -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand
2025/01/22 21:02:26 WARN -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings
2025/01/22 21:02:26 WARN -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign
2025/01/22 21:02:26 WARN -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value
2025/01/22 21:02:26 WARN -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable
2025/01/22 21:02:26 WARN -Wmisleading-indentation -Wundef  -fPIC  -o bcrypt_pbkdf_ext.o -c
2025/01/22 21:02:26 WARN bcrypt_pbkdf_ext.c
2025/01/22 21:02:26 WARN In file included from /usr/include/ruby-3.4.0/ruby/internal/config.h:22,
2025/01/22 21:02:26 WARN                  from /usr/include/ruby-3.4.0/ruby/ruby.h:15,
2025/01/22 21:02:26 WARN                  from /usr/include/ruby-3.4.0/ruby.h:38,
2025/01/22 21:02:26 WARN                  from bcrypt_pbkdf_ext.c:2:
2025/01/22 21:02:26 WARN /usr/include/ruby-3.4.0/aarch64-linux-gnu/ruby/config.h:83:40: fatal error:
2025/01/22 21:02:26 WARN jemalloc/jemalloc.h: No such file or directory
2025/01/22 21:02:26 WARN    83 | #define RUBY_ALTERNATIVE_MALLOC_HEADER <jemalloc/jemalloc.h>
2025/01/22 21:02:26 WARN       |                                        ^
2025/01/22 21:02:26 WARN compilation terminated.
2025/01/22 21:02:26 WARN make: *** [Makefile:251: bcrypt_pbkdf_ext.o] Error 1
2025/01/22 21:02:26 WARN
2025/01/22 21:02:26 WARN make failed, exit code 2
2025/01/22 21:02:26 WARN
```

xnox commented:
ruby-dev needs to depend on jemalloc-dev?

dannf commented:
> ruby-dev needs to depend on jemalloc-dev?

yeah, that oughta do it. i'm running some tests locally & i can push the change

dannf committed:
ruby-3.*-dev: Add runtime-dep on jemalloc-dev

Fix failures like this one when testing ruby3.4-rails-8.0:

 In file included from /usr/include/ruby-3.4.0/ruby/internal/config.h:22,
                  from /usr/include/ruby-3.4.0/ruby/ruby.h:15,
                  from /usr/include/ruby-3.4.0/ruby.h:38,
                  from bcrypt_pbkdf_ext.c:2:
 /usr/include/ruby-3.4.0/aarch64-linux-gnu/ruby/config.h:83:40: fatal error:
 jemalloc/jemalloc.h: No such file or directory
    83 | #define RUBY_ALTERNATIVE_MALLOC_HEADER <jemalloc/jemalloc.h>
       |                                        ^
 compilation terminated.

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

joshrwolf commented:
thanks dan!

@xnox do you mind ✅ now that dan and I have both committed?
